sky/serve/load_balancer.py

```diff
@@ -77,23 +78,85 @@ def _sync_with_controller(self):
+        Returns:
+            The response from the endpoint replica. None if anything goes wrong.
+        """
+        method = request.method
+        headers = {key: value for key, value in request.headers.items()}
+        body = await request.body()
+        path = f'http://{url}{request.url.path}'
+        logger.info(f'Proxy request to {path}')
+        try:
+
+            async def stream_response():
+                async with httpx.AsyncClient() as client:
+                    async with client.stream(method,
+                                             path,
+                                             headers=headers,
+                                             content=body) as response:
+                        response.raise_for_status()
+                        # TODO(tian): Hacky. Investigate a way to not directly
+                        # yielding the response status code and headers.
+                        yield response.status_code
+                        yield dict(response.headers)
+                        try:
+                            async for chunk in response.aiter_bytes():
+                                yield chunk
+                        finally:
+                            await response.aclose()
+
+            content = stream_response()
+            status_code = await content.__anext__()
+            headers = await content.__anext__()
+            return fastapi.responses.StreamingResponse(content=content,
+                                                       status_code=status_code,
+                                                       headers=headers)
+        except (httpx.RequestError, httpx.HTTPStatusError) as e:
+            logger.error(f'Error when proxy request to {path}: '
+                         f'{common_utils.format_exception(e)}')
+            return None
+
+    async def _proxy_with_retries(
+            self, request: fastapi.Request) -> fastapi.responses.Response:
+        """Try to proxy the request to the endpoint replica with retries."""
+        self._request_aggregator.add(request)
+        # TODO(tian): Finetune backoff parameters.
+        backoff = common_utils.Backoff(initial_backoff=1)
+        # SkyServe supports serving on Spot Instances. To avoid preemptions
+        # during request handling, we add a retry here.
+        retry_cnt = 0
+        while True:
+            retry_cnt += 1
+            ready_replica_url = self._load_balancing_policy.select_replica(
+                request)
+            if ready_replica_url is None:
+                raise fastapi.HTTPException(
+                    status_code=503,
+                    detail='No ready replicas. '
+                    'Use "sky serve status [SERVICE_NAME]" '
+                    'to check the replica status.')
+            response = await self._proxy_request_to(ready_replica_url, request)
+            if response is not None:
+                return response
```

Comment:
What will happen if the replica dies when the streaming is not finished? Will we retry here?

Reply:
Currently we dont. One challenge here is to recover the state of streaming, i.e. after a recovery the new replica needs to start streaming from the middle to make the client does not aware that an error has happened. (e.g. the total steaming array is [hello, nice, to, meet, you] and an error happened after [hello, nice]. the new replica needs to stream [to, meet, you] but not the whole array [hello, nice, to, meet, you]) But how could the new replica know where to start? It might be possible to do some special case handling for the LLM workloads (record some information, ...) but it will be really hard to support general streaming workloads.

Reply:
Makes sense. In that case, if an underlying replica is scaled down or preempted, will the client just hang and take a long time to timeout, it might be good to include a TODO here if that is the case to optimize for replica failure?

Reply:
It will immediately closed due to connection lost

```bash
Hello
world!
Nice
to
curl: (18) transfer closed with outstanding read data remaining
```

Reply:
If you decide to store state:
> ... special case handling for the LLM workloads (**record some information**, ...) ...

It would be ideal to turn this off in order to ensure one can boot a zero retention for security.

Reply:
Yeah definitely! We will provide an option to enable this feature or not if we implemented this feature. Thanks for the advise!